### DISCUSSION TITLE
Question about VSG with current saturation

### DISCUSSION THREAD

**Discussion by xiaoheimai95** [CURENT/andes#441] [Q&A]
Dear Dr. Cui and Jinning,
After I modify the code of `REGCV1` so that it can meet the current limitation issue (code is shown below), I test it with ieee39 and kundur cases. And it always raise the problem of initialization and convergence failed. Could you help me to fix this problem?

The error report of kundur case as an example:
```
Suspect initialization issue! Simulation may crash!
       Name        | Var. Value | Eqn. Mismatch
-------------------+------------+--------------
 PIId_xi REGCV1 1  | 0          | 0.654        
 PIId_xi REGCV1 2  | 0          | 0.630        
 PIId_xi REGCV1 3  | 0          | 0.630        
 PIId_xi REGCV1 4  | 0          | 0.630        
 PIIq_xi REGCV1 1  | 0          | -0.099       
 PIIq_xi REGCV1 2  | 0          | -0.205       
 PIIq_xi REGCV1 3  | 0          | -0.209       
 PIIq_xi REGCV1 4  | 0          | -0.095       
 Iref REGCV1 1     | 10.800     | 7.350        
 Iref REGCV1 2     | 10.800     | 7.362        
 Iref REGCV1 3     | 10.800     | 7.376        
 Iref REGCV1 4     | 10.800     | 7.080        
 Idref1_y REGCV1 1 | 0          | 7.268        
 Idref1_y REGCV1 2 | 0          | 7.000        
 Idref1_y REGCV1 3 | 0          | 7.000        
 Idref1_y REGCV1 4 | 0          | 7.000        
 Iqref1_y REGCV1 1 | 0          | -1.095       
 Iqref1_y REGCV1 2 | 0          | -2.280       
 Iqref1_y REGCV1 3 | 0          | -2.324       
 Iqref1_y REGCV1 4 | 0          | -1.061       
 udref REGCV1 1    | 1.024      | 13.058       
 udref REGCV1 2    | 1.051      | 12.549       
 udref REGCV1 3    | 1.052      | 12.548       
 udref REGCV1 4    | 1.024      | 12.576       
 uqref REGCV1 1    | 0.162      | -2.132       
 uqref REGCV1 2    | 0.156      | -4.260       
 uqref REGCV1 3    | 0.156      | -4.338       
 uqref REGCV1 4    | 0.156      | -2.065       
Initialization failed!!
```
```
Jacobian matrix is singular.
Suspect diagonal elements: [101, 102, 103, 104]
NaN found in solution. Convergence is not likely
```

**Comment by jinningwang** [comment]
It seems that the initialization error is caused by the initialization equations. Take a look at this https://github.com/CURENT/andes/discussions/172. 

Here are some other tips that might be helpful:
1. It would be better to create a new model rather than revise the existing model. For example, you can have your revised model named ``VSGOuterPIModel2`` and ``VSGInnerPIModel2`` or any other meaningful name. Then, manufacturing a new model named ``REGCV3`` with the revised PI model. In this way, you can easily compare your model with the existing ``REGCV1``. 
1. During the model development, it can be easier to include only one device of the developed model in the system.
1. After a failed ``TDS.init()``, you can print out the related variables to see if they have a reasonable or expected value.

**Reply by xiaoheimai95** [reply]
Hi Jinning, when I change the name into REGCV3, it report that `<REGCV3> is not an existing model.`.
What should I do to fix it?

**Reply by jinningwang** [reply]
Check this, https://docs.andes.app/en/latest/modeling/examples.html#finalize

**Reply by xiaoheimai95** [reply]
Thx, Jinning!

**Comment by cuihantao** [comment]
Here's a command to output the data used for initialization:
```bash
andes -v 10 run ieee14_linetrip.xlsx -r tds --init &> init_debug.txt
```

The text file logs the initialization sequence, the inputs and outputs of the initialization, etc.

**Reply by xiaoheimai95** [reply]
Dear Dr.Cui, when I use the code in the terminal, it still shows that `REGCV3 is not an exitsting model.` but when I use Pycharm it can link to regcv3 model.

**Reply by jinningwang** [reply]
Personally speaking, I am using VSCode actively. When working in the terminal, you might need to make sure you are using the environment that has ANDES installed in development mode.

**Reply by xiaoheimai95** [reply]
I'm sure it is installed cause when I test other case without REGCV3, it can run smoothly.

**Reply by jinningwang** [reply]
As Dr. Cui commented below, you might have more than one copy installed. You can run ``andes`` in terminal to get the version information to check which version are you using.

**Reply by xiaoheimai95** [reply]
I will recheck later. And to be on the safe side, is it possible to print out the init_debug.txt message via python code?

**Reply by jinningwang** [reply]
If the message does not involve sensitive information of yourself, it's ok to post it here for discussion.

**Reply by cuihantao** [reply]
The messages in `init_debug.txt` are actually the stderr logs. There's currently no API to export the messages via a Python API.

**Comment by cuihantao** [comment]
Likely you have more than one copies of ANDES installed in different environments. The one working is installed in the development mode, but the one not working is installed as a site package.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by xiaoheimai95** [comment]
Dear Dr. Cui and Jinning,
The init_debug.txt is shown. Would you please give me some advice?
[init_debug.txt](https://github.com/CURENT/andes/files/11291668/init_debug.txt)

**Reply by jinningwang** [reply]
Given the following version info, I feel this is not the development mode:
```
    _           _         | Version 1.8.7
   /_\  _ _  __| |___ ___ | Python 3.10.10 on Darwin, 04/21/2023 10:00:34 AM
  / _ \| ' \/ _` / -_|_-< | 
 /_/ \_\_||_\__,_\___/__/ | This program comes with ABSOLUTELY NO WARRANTY.
```

Check this, https://docs.andes.app/en/latest/getting_started/install.html#develop-install

**Reply by xiaoheimai95** [reply]
I will fix that problem later. And how about the initialization failed? How can I fix it?

**Reply by jinningwang** [reply]
Check the equations related to variables in Eqn. mismatch.

**Reply by xiaoheimai95** [reply]
Hi Jinning, I found that the value of `Iref` is equal to the Eqn. Mismatch, but it still has the problem of `Initialization failed`.
![1682067698067(1)](https://user-images.githubusercontent.com/71546641/233594183-e37245f6-b03f-4ad9-9626-dc8f9c7c3fc7.png)

**Reply by jinningwang** [reply]
1. Var. Value is not used to compared with Eqn. Mismatch. Check this https://docs.andes.app/en/latest/modeling/variables.html?highlight=equation#value-and-equation-strings
2. ``AliasAlgeb`` is not used for equations but final output. Check the init message and take a look on the sequence of the variables assignment, also take a look on this, https://docs.andes.app/en/latest/modeling/_generated/andes.core.var.AliasAlgeb.html?highlight=aliasalgeb

**Reply by xiaoheimai95** [reply]
So I have already set `Iref` And I've checked the formula for correctness.

**Reply by jinningwang** [reply]
``e_str`` might be wrong, check other models.

**Reply by cuihantao** [reply]
See https://docs.andes.app/en/latest/modeling/_generated/andes.core.var.Algeb.html#andes-core-var-algeb for `e_str`.

**Reply by xiaoheimai95** [reply]
Thank you very much!
